src/main/java/org/broadinstitute/hellbender/tools/coveragemodel/cachemanager/ImmutableComputableGraph.java

```diff
@@ -0,0 +1,464 @@
+ *
+ * For computable nodes, one must provide the function, a list of immediate parent nodes, a list of tags,
+ * and whether or not the value is to be cached. Computable nodes come in 3 species depending on the way
+ * they are constructed:
+ *
+ *      (1) Caching: these nodes may store the values they evaluate for future lookup
+ *      (2) Non-caching: these nodes are compute-on-demand
+ *      (3) Externally set: these nodes are constructed using a {@code null} evaluation function. These user is
+ *          responsible for evaluating these nodes when required and using {@link #setValue(String, Duplicable)}
+ *          to update them. This class performs the bookkeeping of the status of their values (up to date or
+ *          out of date) based on the provided parents list.
+ *
+ * One require the caching computable nodes to be updated and cached automatically after each mutation of a primitive
+ * or externally set computable node by invoking {@link ImmutableComputableGraphBuilder#enableCacheAutoUpdate()}.
+ * This is feature, however, is NOT recommended as one may not need all the caches to be up-to-date at all times
+ * (see below for updating caches selectively). This class with throw an {@link IllegalStateException} if an old
+ * cache is invoked in order to notify the user to update the cache manually.
+ *
+ * Updating caches:
+ * ================
+ *
+ * If cache auto update is not enabled, the user is responsible for updating the caches by calling
+ * either {@link #updateAllCaches()}, {@link #updateCachesForNode(String)}, or {@link #updateCachesForTag(String)}.
+ *
+ * Querying:
+ * =========
+ *
+ * The graph is queried either by calling {@link #getValueDirect(String)} or
+ * {@link #getValueWithRequiredEvaluations(String)}. The former only fetches the values and throws an exception
+ * if the some of the required caches are out of date, or a non-caching computable node is encountered.
+ * The latter performs the required evaluations along the way.
+ *
+ * IMPORTANT NOTE: the queried values are return by reference. It is the user's responsibility not to mutate
+ * them. Otherwise, the functional structure will be broken.
+ *
+ * Mutation:
+ * =========
+ *
+ * The primitive values can be mutated by calling {@link ImmutableComputableGraph#setValue(String, Duplicable)}.
+ * The mutations do not occur in place; rather, a new instance of {@link ImmutableComputableGraph} is created along
+ * with new instances for the updated nodes. If cache auto update is enabled, the affected nodes will be evaluated.
+ * Otherwise, only the cache status will go out of date. Unchanged nodes are passed as reference to the new instance.
+ * JVM's garbage collector will free up the memory for old cached nodes in sequential computations.
+ *
+ * @author Mehrtash Babadi &lt;mehrtash@broadinstitute.org&gt;
+ */
+public class ImmutableComputableGraph implements Serializable {
```

Comment:
I meant it from the design perspective. If class is not final it can be extended later in a way that breaks the immutability contract. I don't think it matters too much in this case though..